### PR TITLE
Using ParserInt from muparser when '0x' is given to support hex math.

### PR DESCRIPTION
**Pull request by m-radomski**
A basic way of using muparser libs ability to work on hex encoded expressions.

### CODE REVIEW

File: calculator/src/extension.cpp [hunk 1]
```diff
@@ -31,6 +32,7 @@ class Calculator::Private
 public:
     QPointer<ConfigWidget> widget;
     std::unique_ptr<mu::Parser> parser;
+    std::unique_ptr<mu::ParserInt> iparser;
```
Comment (ManuelSchneid3r) [inline calculator/src/extension.cpp:35]: arithmetic hex operations are a extreme edge case. conversion fine, but virtually nobody will calculate hex values. I'd like to see this as an option.  besides, do you know the memory impact of such a parser?

**Review by ManuelSchneid3r — CHANGES_REQUESTED** [review]

File: calculator/src/extension.cpp [hunk 2]
```diff
@@ -86,8 +93,17 @@ QWidget *Calculator::Extension::widget(QWidget *parent) {
 /** ***************************************************************************/
 void Calculator::Extension::handleQuery(Core::Query * query) const {
 
+    bool isHex = query->string().toStdString().find("0x") != std::string::npos;
```
[the remark this answers is not in the document]
Comment (m-radomski) [reply calculator/src/extension.cpp]: I agree with that. Changed it with 188ff6678c0a202fa80916f87e8bb3ff853774e5.

File: calculator/src/extension.cpp (hunk 1 above)
Comment (m-radomski) [reply calculator/src/extension.cpp:36]: I don't know if I understood you, but I think changing it to be a NULL pointer and then actually creating it when someone requests "0x".

Comment (m-radomski) [reply calculator/src/extension.cpp:38]: I solved it using a check box in settings->extensions, changes in efcdf21e9fc7cdb086742e320c4802f259597137.

File: calculator/src/extension.cpp [hunk 3]
```diff
@@ -86,8 +109,24 @@ QWidget *Calculator::Extension::widget(QWidget *parent) {
...
     try {
-        d->parser->SetExpr(query->string().toStdString());
+        if(hexParse)
+        {
+            d->iparser->SetExpr(query->string().toStdString());
+        }
+        else
+        {
+            d->parser->SetExpr(query->string().toStdString());
+        }
```
Comment (ManuelSchneid3r) [inline calculator/src/extension.cpp:129]: this block is weird

```cpp
if(d->iparser && query->string().contains("0x")) 
    d->iparser->SetExpr(query->string().toStdString());
else 
    d->parser->SetExpr(query->string().toStdString());
```

**Review by ManuelSchneid3r — CHANGES_REQUESTED** [review]

File: calculator/src/extension.cpp (hunk 3 above)
Comment (dshoreman) [reply calculator/src/extension.cpp:129]: I'd actually keep the curlies, but have them on the same line as if/else. That way it looks nicer, still allows editors to highlight the matching brace, *and* any future additions to the code inside them won't litter diffs just to add `{}` around it :wink: 

```cpp
if (d->iparser && query->string().contains("0x")) {
    // ...
} else {
    // ...
}
```

File: calculator/src/extension.cpp [hunk 4]
```diff
@@ -77,6 +84,22 @@ QWidget *Calculator::Extension::widget(QWidget *parent) {
             d->locale.setNumberOptions( (checked) ? d->locale.numberOptions() & ~QLocale::OmitGroupSeparator
                                                   : d->locale.numberOptions() | QLocale::OmitGroupSeparator );
         });
+
+        d->widget->ui.checkBox_hexparsing->setChecked(!(d->locale.numberOptions() & QLocale::OmitGroupSeparator));
+        connect(d->widget->ui.checkBox_hexparsing, &QCheckBox::toggled, [this](bool checked){
+                settings().setValue(CFG_HEXP, checked);
+                if(checked)
+                {
```
Comment (dshoreman) [inline calculator/src/extension.cpp:92]:
```suggestion
                settings().setValue(CFG_HEXP, checked);

                if (checked) {
```

File: calculator/src/extension.cpp [hunk 5]
```diff
@@ -77,6 +84,22 @@ QWidget *Calculator::Extension::widget(QWidget *parent) {
...
+                    d->iparser.reset(new mu::ParserInt);
+                    d->iparser->SetDecSep(d->locale.decimalPoint().toLatin1());
+                    d->iparser->SetThousandsSep(d->locale.groupSeparator().toLatin1());
+                    d->iparser->SetArgSep(';');
+                }
+                else
+                {
```
Comment (dshoreman) [inline calculator/src/extension.cpp:99]:
```suggestion
                } else {
```

File: calculator/src/extension.cpp [hunk 6]
```diff
@@ -86,8 +109,24 @@ QWidget *Calculator::Extension::widget(QWidget *parent) {
 /** ***************************************************************************/
 void Calculator::Extension::handleQuery(Core::Query * query) const {
 
+    bool hexParse = settings().value(CFG_HEXP, CFG_HEXP_DEF).toBool();
+
+    if(hexParse)
+    {
```
Comment (dshoreman) [inline calculator/src/extension.cpp:115]:
```suggestion
    if (hexParse) {
```

File: calculator/src/extension.cpp (hunk 3 above)
Comment (dshoreman) [inline calculator/src/extension.cpp:129]:
```suggestion
        if (hexParse) {
            d->iparser->SetExpr(query->string().toStdString());
        } else {
            d->parser->SetExpr(query->string().toStdString());
        }
```

File: calculator/src/extension.cpp [hunk 7]
```diff
@@ -96,7 +135,14 @@ void Calculator::Extension::handleQuery(Core::Query * query) const {
 
     // http://beltoforion.de/article.php?a=muparser&p=errorhandling
     try {
-        result = d->parser->Eval();
+        if(hexParse)
+        {
+            result = d->iparser->Eval();
+        }
+        else
+        {
```
Comment (dshoreman) [inline calculator/src/extension.cpp:143]:
```suggestion
        if (hexParse) {
            result = d->iparser->Eval();
        } else {
```

**Review by dshoreman — CHANGES_REQUESTED** [review]
:+1:  Love the idea of adding muparser, it can do "logical" calculations like "135.70781102 - 69%" too which aren't supported currently.

@ManuelSchneid3r I've suggested some changes to cleanup the if statements for now, can test locally once the file reading / pointer bit is resolved.

**Comment by m-radomski** [comment]
71815c9b0e4a5eb36401b84ad1dd5dabe45b2b94 addressed both replies, now checking if pointer is non-NULL instead of going to `settings`. Chose https://github.com/albertlauncher/plugins/pull/42#discussion_r539033634 approach with using curlies that keep with project's coding style.

**Comment by ManuelSchneid3r** [comment]
Sorry this takes too long. I merged and added some changes on top. Ty

**Pull request closed without merging**